### ISSUE TITLE
Investigate new mongodb scala driver

### ISSUE THREAD

**Issue by scullxbones** [scullxbones/akka-persistence-mongo#55]
[New driver](https://github.com/mongodb/mongo-scala-driver) is possibly a casbah replacement, would need to run 3 drivers for a while and see the uptake.

RxScala and reactive streams would make some facets simpler (e.g. ReadJournal)

**Comment by scullxbones** [comment]
[Driver released 1.0](http://rosslawley.co.uk/mongodb-scala-driver-released/)

**Comment by scullxbones** [comment]
It looks like `casbah` will be unsupported for mongo `3.2`+

**Comment by matheuslima** [comment]
I think this driver is a good option to the plugin
https://mongodb.github.io/mongo-java-driver-reactivestreams/
This driver expose mongo operations as reactive streams publishers

**Comment by filosganga** [comment]
The 1.2.1 version has been released:
https://github.com/mongodb/mongo-scala-driver/releases/tag/r1.2.1

**Comment by scullxbones** [comment]
@filosganga there are quite a few users that still use casbah, including me 😄  ... it would be nice though to have some kind of quantitative metrics on what drivers are used.  The reality is that casbah is easier to support than rxmongo by far.  It's extremely stable.

I'm thinking for this ticket I'd like an additional 3rd driver.

**Comment by filosganga** [comment]
@scullxbones Makes sense, I have just realised that rxmongo is not widely adopted and casbah is still the most popular so far.

**Comment by wesplunk** [comment]
Is this still being pursued in light of the EOL of casbah?

**Comment by scullxbones** [comment]
Hi @wesplunk - there's no active effort on it at the moment, but it's definitely getting more important.

**Comment by bykka** [comment]
Hi, another idea! I see that RxMongo already converts reactivemongo to akka streams. What if to try to use alpakka? Alpakka also converts everything to akka streams. And in case of mongo knows how to read and write using last mongo driver for scala. Maybe it will simplify integration with other sources, for example ElasticSearch.

**Comment by gbrd** [comment]
I started naively to implement some methods to try to understand where difficulty is.
https://github.com/gbrd/akka-persistence-mongo/tree/feature-new-scala-driver
To be continued, don't know the best way to collaborate on this.

**Comment by scullxbones** [comment]
It's a large effort, no doubt.  I don't think it's an especially difficult effort, but there is a significant footprint to fill:

* extension implementation
* write-side journaller, snapshotter
* read-side queries
* serializers
* miscellaneous parts I haven't thought of

There's a bit of reusability especially on the read sides, and serializers, but write side is older and not as abstracted.  

I had thought with this story I'd try to tackle doing a better job of abstraction across all drivers, possibly using a free monad approach where the shared code specifies the operations and the interpreter is the driver-specific plugin.

Those are a couple of the bigger reasons this story has not progressed, although it is getting to be quite important to support future evolution.  Mongo server has kept backward compatible so far with the fairly small portion of the feature set that this plugin uses, so have been fortunate there at least.

**Comment by scullxbones** [comment]
I've started work on this, it'll be available in a future release.  Hopefully in a couple of weeks if I can keep current velocity.
